File: Problems/0283.Move-Zeroes/MoveZeroes.go
package leetcode

// Runtime: 4 ms, faster than 92.97% of Go online submissions for Move Zeroes.
func moveZeroes(nums []int) {
	length := len(nums)
	pt1 := 0
	pt2 := 1
	for true {
		for pt1 < length && nums[pt1] != 0 {
			pt1++
		}
		for pt2 < length && (pt2 < pt1 || nums[pt2] == 0) {
			pt2++
		}
		if pt1 == length || pt2 == length {
			break
		}
		nums[pt1], nums[pt2] = nums[pt2], nums[pt1]
	}
}

File: leetcode/0283.Move-Zeroes/MoveZeroes.py
from typing import List

class Solution:

    # Runtime: 44 ms, faster than 95.39% of Python3 online submissions for Move Zeroes.
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        pt1 = 0
        pt2 = 1
        length = len(nums)
        while True:
            while pt1 < length and nums[pt1] != 0:
                pt1 += 1
            while pt2 < length and (pt2 < pt1 or nums[pt2] == 0):
                pt2 += 1
            if pt1 == length or pt2 == length:
                break
            nums[pt1], nums[pt2] = nums[pt2], nums[pt1]